### PR TITLE
Expand embedded linker discovery and improve error messages.

### PR DESCRIPTION
**Pull request by ScottTodd**
This path didn't _quite_ work out of the box for development builds. This expands the search space to make automatic discovery more likely.

The LLVM backend's embedded linker requires `lld` to be discoverable and we make an attempt to build it from source:
https://github.com/google/iree/blob/222828129144c8bfb603b4647aa969afb64a5a45/iree/tools/CMakeLists.txt#L395-L401

However, the `lld` binary ends up in `iree-build\third_party\llvm-project\llvm\bin` for builds (`install` does put `lld` next to the other IREE tools). This now searches that path in the build tree and includes more information in error messages:
```
Searching environment for one of these tools: [iree-lld.exe, lld.exe, ld.lld.exe, lld-link.exe]
Searching for tool at path 'D:\dev\projects\iree\..\iree-build\iree\tools'
Searching for tool at path 'D:\dev\projects\iree\..\iree-build\iree\tools/../../third_party/llvm-project/llvm/bin/'

(success)
Found tool in build tree at path "D:\dev\projects\iree-build\third_party\llvm-project\llvm\bin\lld.exe"

(failure)
error: embedded linker tool (typically `lld`) not found after searching:
  * -iree-llvm-embedded-linker-path= flag
  * IREE_LLVMAOT_EMBEDDED_LINKER_PATH environment variable
  * common locations at relative file paths
  * system PATH
Run with -debug-only=llvmaot-linker for search details
```

We do also have `iree-lld` now, but that lives in [llvm-external-projects/iree-compiler-api/python/](https://github.com/google/iree/blob/main/llvm-external-projects/iree-compiler-api/python/), not [iree/tools/](https://github.com/google/iree/tree/main/iree/tools/).

A workaround we _don't_ want to encourage is to install `lld` directly at the system level. Using the built-from-source binary has less risk of compatibility issues.

### CODE REVIEW

**Comment by benvanik** [comment]
Not so sure about this - we explicitly don't want people using system linkers, and if they can't find the lld we build then the solution is to build it, not use the wrong one :)

**Comment by benvanik** [comment]
(there are a bunch of reasons this is the case for the embedded linker path, namely that we need to ensure the lld has the target architectures we need as is specified in the cmake configuration - when we enable LTO we need to ensure the lld version is bitcode-compatible with the bitcode we produce using the submodule llvm version, which with an aged system lld won't always be the case)

**Comment by ScottTodd** [comment]
Fair points. Let's see how we can discover the `lld` or `iree-lld` we do build then.

**Comment by benvanik** [comment]
We could update findToolInEnvironment to look in some well known paths (bazel/cmake/python/etc); it's already searching around for things: https://github.com/google/iree/blob/981f48768a5a27baac818b8afa506abaf4ba75ad/iree/compiler/Dialect/HAL/Target/LLVM/internal/EmbeddedLinkerTool.cpp#L58

If we consistently call it iree-lld we wouldn't even have to feel bad about looking around - we shouldn't hit name conflicts anywhere like that.

File: iree/compiler/Dialect/HAL/Target/LLVM/LinkerTool.cpp [hunk 1]
```diff
@@ -165,6 +165,28 @@ std::string LinkerTool::findToolInEnvironment(
...
+  if (!mainExecutable.empty()) {
+    llvm::sys::path::remove_filename(mainExecutable);
+
+    for (auto toolName : toolNames) {
+      std::string normalizedToolName = normalizeToolNameForPlatform(toolName);
+
+      SmallString<256> pathStorage;
+      llvm::sys::path::append(pathStorage, std::string(mainExecutable),
+                              "../../third_party/llvm-project/llvm/bin/",
+                              normalizedToolName);
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/LLVM/LinkerTool.cpp:180]: Ok, this path searching in the build directory _seems_ to work on my Windows machine at least... but it isn't pretty :P

**Comment by KoolJBlack** [comment]
Re Ben's comment: whats the reason we don't want people using system linkers? I ran into this issue due to the recent requirement of `lld` not being widely published and the error messages don't make that clear. I was able to resolve it by `apt-get` installing the linker in question. Are there esoteric compatibility issues I'll run into?

File: iree/compiler/Dialect/HAL/Target/LLVM/LinkerTool.cpp (hunk 1 above)
Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/Target/LLVM/LinkerTool.cpp:180]: (Cleaning this up then I'll send for review again)

**Comment by ScottTodd** [comment]
> Re Ben's comment: whats the reason we don't want people using system linkers? I ran into this issue due to the recent requirement of `lld` not being widely published and the error messages don't make that clear. I was able to resolve it by `apt-get` installing the linker in question. Are there esoteric compatibility issues I'll run into?

See Ben's second comment. Yes, compatibility issues are possible.

**Comment by ScottTodd** [comment]
PTAL. I cleaned up the search code and added more to the error messages. See the PR description for a sample of the output now.

File: iree/compiler/Dialect/HAL/Target/LLVM/internal/EmbeddedLinkerTool.cpp [hunk 2]
```diff
@@ -88,8 +93,11 @@ class EmbeddedLinkerTool : public LinkerTool {
     }
     artifacts.libraryFile.close();
 
+    std::string embeddedToolPath = getEmbeddedToolPath();
+    if (embeddedToolPath.empty()) return llvm::None;
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/LLVM/internal/EmbeddedLinkerTool.cpp:97]: Together with the other changes, this early exit changes the error message from something like

```
LLD (ld.lld) not found on path; specify with the IREE_LLVMAOT_EMBEDDED_LINKER_PATH environment variable or -iree-llvm-embedded-linker-path=
sh: line 1: -flavor: command not found
Linking failed; escaped command line returned exit code 32512:

 -flavor gnu -o /tmp/simple_mul_dispatch_0-ed0c39.so -nostdlib -static -shared --gc-sections -z now -z relro --discard-all --hash-style=sysv /tmp/simple_mul_dispatch_0-ed0c39.o /tmp/simple_mul_dispatch_0-ed0c39.librt.o /tmp/simple_mul_dispatch_0-ed0c39.librt64.o

/usr/local/google/home/kooljblack/Code/iree-data/simple_mul.mlir:3:8: error: failed to link executable and generate target dylib using linker toolchain 
  %0 = "arith.mulf"(%arg0, %arg1) {name = "mul.1"} : (tensor<4xf32>, tensor<4xf32>) -> tensor<4xf32>
       ^
```

to

```
error: required embedded linker tool (typically `lld`) not found after searching:
  * -iree-llvm-embedded-linker-path= flag
  * IREE_LLVMAOT_EMBEDDED_LINKER_PATH environment variable
  * common locations at relative file paths
  * system PATH
Run with -debug-only=llvmaot-linker for search details
D:\dev\projects\iree/iree/samples/models/simple_abs.mlir:2:13: error: failed to link executable and generate target dylib (check above for more specific error messages)
  %result = math.abs %input : tensor<f32>
            ^
```

(` -flavor gnu -o ...` doesn't mean much without knowing that there should be a `lld` executable at the start of that command :P)

Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/Target/LLVM/internal/EmbeddedLinkerTool.cpp:97]: Note that this still checks for the linker tool after other parts of compilation have ran, and it still outputs the full "`note: see current operation: "hal.executable.variant"()` {giant MLIR blob here}", so users/developers will still have to scroll up in the error output to see this :/

**Review by benvanik — APPROVED** [review]

**Pull request merged**